Make EnumTypeAdapter friendly with obfuscation

ganadist opened the pull request:
When enum value was obfuscated by proguard, EnumTypeAdapter raise NoSuchFieldException, even if apply SerializedName annotation.
Because EnumTypeAdapter cannot find obfuscated enum constant field with its name.
But without this workaround, there is no way to obfuscate enum name.

JakeWharton commented:
Test case?

ganadist commented:
I tried to integrate proguard obfuscation on testcase like this.

https://github.com/ganadist/gson/commit/251ec476aa49fbb4653b164c67d22e17445625c4

And intent of this mvn rule are this order.
* copy-resources (pre-obfuscate-class)
* proguard (default)
* copy-resources (post-obfuscate-class)

But order of plugin execution does not work as intended.
* copy-resources (pre-obfuscate-class)
* copy-resources (post-obfuscate-class)
* proguard (default)

Can I get any idea?

JakeWharton commented:
You don't need to integrate ProGuard. Just specify a different name and
ensure it's honored in the JSON output.
On Fri, Mar 22, 2019 at 3:37 PM YOUNG HO CHA (aka ganachoco) <
notifications@github.com> wrote:

> I tried to integrate proguard obfuscation on testcase like this.
>
> ganadist@251ec47
> <https://github.com/ganadist/gson/commit/251ec476aa49fbb4653b164c67d22e17445625c4>
>
> And intent of this mvn rule are this order.
>
>    - copy-resources (pre-obfuscate-class)
>    - proguard (default)
>    - copy-resources (post-obfuscate-class)
>
> But order of plugin execution does not work as intended.
>
>    - copy-resources (pre-obfuscate-class)
>    - copy-resources (post-obfuscate-class)
>    - proguard (default)
>
> Can I get any idea?
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/google/gson/pull/1495#issuecomment-475754232>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAEEEabvLIMEqbJOJAtqlKXYhGQvUzj4ks5vZTEVgaJpZM4cDGDK>
> .
>

ganadist commented:
> Just specify a different name and ensure it's honored in the JSON output.

Gson has already [com.google.gson.functional.testEnumCaseMapping()](https://github.com/google/gson/blob/master/gson/src/test/java/com/google/gson/functional/EnumTest.java#L147) testcase.

ganadist commented:
I updated first conversation of this PR. because it could be misleading.

inder123 commented:
I ran into this problem as well. This is a good fix. Let's add a testcase.

ganadist commented:
To verify this change, enum class in testcase must be obfuscated by proguard.
And I updated testcase on https://github.com/ganadist/gson/commit/1b65e3aae9f549c7a6adaf4dc533263cb7e1be4c , added to check enum class is obfuscated.

But it still has a problem to integrate proguard. :(

ganadist commented:
Proguard integration on Maven testcase is pretty dirty, but.
```
Running com.google.gson.functional.EnumWithObfuscatedTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.001 sec
```

``` 
$ cat gson/target/proguard_map.txt
com.google.gson.functional.EnumWithObfuscatedTest -> com.google.gson.functional.EnumWithObfuscatedTest:
    com.google.gson.Gson gson -> a
    void <init>() -> <init>
    void setUp() -> setUp
    void testEnumClassWithObfuscated() -> testEnumClassWithObfuscated
com.google.gson.functional.EnumWithObfuscatedTest$Gender -> com.google.gson.functional.EnumWithObfuscatedTest$Gender:
    com.google.gson.functional.EnumWithObfuscatedTest$Gender MALE -> a
    com.google.gson.functional.EnumWithObfuscatedTest$Gender FEMALE -> b
    com.google.gson.functional.EnumWithObfuscatedTest$Gender[] $VALUES -> c
    com.google.gson.functional.EnumWithObfuscatedTest$Gender[] values() -> values
    com.google.gson.functional.EnumWithObfuscatedTest$Gender valueOf(java.lang.String) -> valueOf
    void <init>(java.lang.String,int) -> <init>
    void <clinit>() -> <clinit>
```

Marcono1234 commented on gson/src/main/java/com/google/gson/internal/bind/TypeAdapters.java:
```diff
@@ -788,7 +794,11 @@ public EnumTypeAdapter(Class<T> classOfT) {
+        throw new AssertionError(e);
+      } catch (NullPointerException e) {
+        throw new AssertionError(e);
+      } catch (ExceptionInInitializerError e) {
```
Why catch this?

ganadist replied:
For https://docs.oracle.com/javase/7/docs/api/java/lang/reflect/Field.html#get(java.lang.Object)

Marcono1234 replied:
Yes, but only because the method can throw the error does not mean that it should be caught here. I assume if that error actually occurred it could also happen in the non-Proguard specific code and there it is not caught either. So in my opinion it should not be caught here.

(Though note that I am not a maintainer of this project)

ganadist replied:
Fixed at eba1415f1c20a225e528aac83913c3a09962c282

Marcono1234 commented on gson/src/main/java/com/google/gson/internal/bind/TypeAdapters.java:
```diff
@@ -788,7 +794,11 @@ public EnumTypeAdapter(Class<T> classOfT) {
-      } catch (NoSuchFieldException e) {
+      } catch (IllegalAccessException e) {
+        throw new AssertionError(e);
+      } catch (NullPointerException e) {
```
Why catch this?

Marcono1234 replied:
> `NullPointerException` - if the specified object is null and the field is an instance field.

But the enum constants are `static` fields so this should not be possible.

ganadist replied:
Fixed at eba1415f1c20a225e528aac83913c3a09962c282

Marcono1234 commented on gson/src/main/java/com/google/gson/internal/bind/TypeAdapters.java:
```diff
@@ -776,9 +777,14 @@ public void write(JsonWriter out, Locale value) throws IOException {
+          if (!field.isEnumConstant()) {
+            continue;
+          }
+          field.setAccessible(true);
```
Why make the field accessible? Enum element fields are public so making them accessible should not be necessary.

ganadist replied:
Without this line, unittests failed with following error.
```
Failed tests:   testCollectionOfEnumsDeserialization(com.google.gson.functional.EnumTest): AssertionError (GSON 2.8.6-SNAPSHOT): java.lang.IllegalAccessException: class com.google.gson.internal.bind.TypeAdapters$EnumTypeAdapter cannot access a member of class com.google.gson.functional.EnumTest$MyEnum with modifiers "public static final"
  testTopLevelEnumDeserialization(com.google.gson.functional.EnumTest): AssertionError (GSON 2.8.6-SNAPSHOT): java.lang.IllegalAccessException: class com.google.gson.internal.bind.TypeAdapters$EnumTypeAdapter cannot access a member of class com.google.gson.functional.EnumTest$MyEnum with modifiers "public static final"
  testClassWithEnumFieldSerialization(com.google.gson.functional.EnumTest): java.lang.IllegalAccessException: class com.google.gson.internal.bind.TypeAdapters$EnumTypeAdapter cannot access a member of class com.google.gson.functional.EnumTest$MyEnum with modifiers "public static final"
  testClassWithEnumFieldDeserialization(com.google.gson.functional.EnumTest): AssertionError (GSON 2.8.6-SNAPSHOT): java.lang.IllegalAccessException: class com.google.gson.internal.bind.TypeAdapters$EnumTypeAdapter cannot access a member of class com.google.gson.functional.EnumTest$MyEnum with modifiers "public static final"
  testCollectionOfEnumsSerialization(com.google.gson.functional.EnumTest): AssertionError (GSON 2.8.6-SNAPSHOT): java.lang.IllegalAccessException: class com.google.gson.internal.bind.TypeAdapters$EnumTypeAdapter cannot access a member of class com.google.gson.functional.EnumTest$MyEnum with modifiers "public static final"
  testTopLevelEnumSerialization(com.google.gson.functional.EnumTest): java.lang.IllegalAccessException: class com.google.gson.internal.bind.TypeAdapters$EnumTypeAdapter cannot access a member of class com.google.gson.functional.EnumTest$MyEnum with modifiers "public static final"

```

eamonnmcmanus replied:
Enum constants are public, but the enum class itself isn't necessarily. If it isn't, then you'll get `IllegalAccessException`.

My concern here is that if there is a SecurityManager, the `field.setAccessible(true)` call might fail, whereas the code using `classOfT.getEnumConstants()` would have worked. So we'll have fixed one problem (getting enum constants in ProGuarded code) but introduced another (getting enum constants when there is a SecurityManager).

I think the code here should read something like this:

```java
for (final Field field : classOfT.getDeclaredFields()) {
  if (!field.isEnumConstant()) {
    continue;
  }
  AccessController.doPrivileged(
      new PrivilegedAction<Void>() {
        @Override Void run() {
          field.setAccessible(true);
          return null;
        }
      });
  @SuppressWarnings("unchecked")
  T constant = (T) field.get(null);
```

ganadist replied:
Fixed at 59a8aedb37ae35d4d2a4306c92efd463fffaaf69

Marcono1234 commented:
This pull request would fix #924 and #1776

Marcono1234 commented on gson/src/main/java/com/google/gson/internal/bind/TypeAdapters.java:
```diff
@@ -776,9 +777,14 @@ public void write(JsonWriter out, Locale value) throws IOException {
+            continue;
+          }
+          field.setAccessible(true);
+          T constant = (T)(field.get(null));
```
Please add a `@SuppressWarnings("unchecked")` for this variable.

ganadist replied:
Fixed at ddee26e283d088fea1ec2807dfb258e9f36b2497

Marcono1234 commented on gson/src/test/java/com/google/gson/functional/EnumWithObfuscatedTest.java:
```diff
@@ -0,0 +1,60 @@
+
+package com.google.gson.functional;
+
+import java.lang.reflect.Field;
```
This import is unused

ganadist replied:
Fixed at 7dbfce9d9633c65c16f595e9c779c80084ddf894

Marcono1234 commented on gson/testcases-proguard.conf:
```diff
@@ -0,0 +1,20 @@
+# Options from Android Gradle Plugins
```
Would it be possible to move this to `gson/src/test/resources`? I think that might be better than having it directly in the `gson` folder.

ganadist replied:
Fixed at a25f96b8936625ff4856495e8f8dfc63205bcbd7

eamonnmcmanus commented on gson/src/test/java/com/google/gson/functional/EnumWithObfuscatedTest.java:
```diff
@@ -0,0 +1,58 @@
+
+  public void testEnumClassWithObfuscated() {
+    for (Gender enumConstant: Gender.class.getEnumConstants()) {
+        try {
```
Would you mind fixing the indentation on lines 48–50 so it uses the same two-space indentation as the rest of the code?

ganadist replied:
Fixed at d8c5fcf00bdd3170365e92e43880021031c7d005

eamonnmcmanus reviewed (changes requested):
Thanks for putting this change together, especially the tests! Sorry it has taken so long for us to get to it.

There are just a couple of changes that I think would make sense.

I'm going to test this change against Google's many internal uses of Gson and if it comes through clean we should be good to go.

eamonnmcmanus reviewed (approved):
Thanks for the quick update!
